native/dlasq2.go

```diff
@@ -0,0 +1,330 @@
+		} else if z[2] > z[0] {
+			z[0], z[2] = z[2], z[0]
+		}
+		z[4] = z[0] + z[1] + z[2]
+		if z[1] > z[2]*tol2 {
+			t := 0.5 * (z[0] - z[2] + z[1])
+			s := z[2] * (z[1] / t)
+			if s <= t {
+				s = z[2] * (z[1] / (t * (1 + math.Sqrt(1+s/t))))
+			} else {
+				s = z[2] * (z[1] / (t + math.Sqrt(t)*math.Sqrt(t+s)))
+			}
+			t = z[0] + s + z[1]
+			z[2] *= z[0] / t
+			z[0] = t
+		}
+		z[1] = z[2]
+		z[5] = z[1] + z[0]
+		return info
+	}
+	z[2*n-1] = 0
+	emin := z[1]
+	var d, e, qmax, zmax float64
+	var i1, n1 int
+	for k := 0; k < 2*(n-1); k += 2 {
+		if z[k] < 0 {
+			panic("lapack: bad z value")
+		} else if z[k+1] < 0 {
+			panic("lapack: bad z value")
+		}
+		d += z[k]
+		e += z[k+1]
+		qmax = math.Max(qmax, z[k])
+		emin = math.Min(emin, z[k+1])
+		zmax = math.Max(math.Max(qmax, zmax), z[k+1])
+	}
+	if z[2*(n-1)] < 0 {
+		panic("lapack: bad z value")
+	}
+	d += z[2*(n-1)]
+	qmax = math.Max(qmax, z[2*(n-1)])
+	zmax = math.Max(qmax, zmax)
+	if e == 0 {
+		for k := 1; k < n; k++ {
+			z[k] = z[2*k]
+		}
+		impl.Dlasrt(lapack.SortDecreasing, n, z)
+		z[2*(n-1)] = d
+		return info
+	}
+	trace := d + e
+	if trace == 0 {
+		z[2*(n-1)] = 0
+		return info
+	}
+
+	for k := 2 * n; k >= 2; k -= 2 {
+		z[2*k-1] = 0
+		z[2*k-2] = z[k-1]
+		z[2*k-3] = 0
+		z[2*k-4] = z[k-2]
+	}
+	i0 := 0
+	n0 := n - 1
+
+	if cbias*z[4*i0] < z[4*n0] {
+		ipn4Out := 4 * (i0 + n0 + 2)
+		for i4loop := 4 * (i0 + 1); i4loop <= 2*(i0+n0+1); i4loop += 4 {
+			i4 := i4loop - 1
+			ipn4 := ipn4Out - 1
+			z[i4-3], z[ipn4-i4-4] = z[ipn4-i4-4], z[i4-3]
+			z[i4-1], z[ipn4-i4-6] = z[ipn4-i4-6], z[i4-1]
+		}
+	}
+
+	pp := 0
+	for k := 0; k < 2; k++ {
+		d = z[4*n0+pp]
+		for i4loop := 4*n0 + pp; i4loop >= 4*(i0+1)+pp; i4loop -= 4 {
+			i4 := i4loop - 1
+			if z[i4-1] <= tol2*d {
+				z[i4-1] = math.Copysign(0, -1)
+				d = z[i4-3]
+			} else {
+				d = z[i4-3] * (d / (d + z[i4-1]))
+			}
+		}
+		emin = z[4*(i0+1)+pp]
+		d = z[4*i0+pp]
+		for i4loop := 4*(i0+1) + pp; i4loop <= 4*n0+pp; i4loop += 4 {
+			i4 := i4loop - 1
+			z[i4-2*pp-2] = d + z[i4-1]
+			if z[i4-1] <= tol2*d {
+				z[i4-1] = math.Copysign(0, -1)
+				z[i4-2*pp-2] = d
+				z[i4-2*pp] = 0
+				d = z[i4+1]
+			} else if safmin*z[i4+1] < z[i4-2*pp-2] && safmin*z[i4-2*pp-2] < z[i4+1] {
+				tmp := z[i4+1] / z[i4-2*pp-2]
+				z[i4-2*pp] = z[i4-1] * tmp
+				d *= tmp
+			} else {
+				z[i4-2*pp] = z[i4+1] * (z[i4-1] / z[i4-2*pp-2])
+				d = z[i4+1] * (d / z[i4-2*pp-2])
+			}
+			emin = math.Min(emin, z[i4-2*pp])
+		}
+		z[4*(n0+1)-pp-3] = d
+		qmax = z[4*(i0+1)-pp-3]
+		for i4loop := 4*(i0+1) - pp + 2; i4loop <= 4*(n0+1)+pp-2; i4loop += 4 {
+			i4 := i4loop - 1
+			qmax = math.Max(qmax, z[i4])
+		}
+		pp = 1 - pp
+	}
+
+	var ttype int
+	var dmin1, dmin2, dn, dn1, dn2, g, tau float64
+	var tempq float64
+	iter := 2
+	var nFail int
+	nDiv := 2 * (n0 - i0)
+	var i4 int
+	for iwhila := 1; iwhila <= n+1; iwhila++ {
+		if n0 < 0 {
+			goto OneSeventy
+		}
+		var desig float64
+		var sigma float64
+		if n0 != n-1 {
```

Comment:
The actual code description is "Local Variables: I0:N0 defines a current unreduced segment of Z". The unreduced part is useful, but otherwise it's incomplete. It doesn't really help here either -- why is sigma taking values from z[4*n]? The description of z only talks about the z[2*n] elements. 

I'll add in the comments. I'm just saying I'm with you.